app/src/main/java/fr/neamar/kiss/cache/MemoryCacheHelper.java

```diff
@@ -0,0 +1,165 @@
+package fr.neamar.kiss.cache;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.graphics.drawable.Drawable;
+import android.os.AsyncTask;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.lang.ref.WeakReference;
+import java.util.HashMap;
+
+import fr.neamar.kiss.KissApplication;
+import fr.neamar.kiss.utils.UserHandle;
+
+public class MemoryCacheHelper {
+
+    private static final HashMap<AppIconHandle, Drawable> sAppIconCache = new HashMap<>();
+
+    /**
+     * If the app icon is not found in the cache we load it. Else return cache value. Asynchronous function.
+     *
+     * @param context    context to use
+     * @param className  app
+     * @param userHandle android.os.UserHandle
+     * @param callback   this will be called when we know what the drawable is
+     */
+    public static <T> void getAppIconDrawable(@NonNull Context context, ComponentName className, UserHandle userHandle, @Nullable OnDrawableReady<T> callback) {
+        AppIconHandle handle = new AppIconHandle(className, userHandle);
+        Drawable drawable;
+        synchronized (sAppIconCache) {
+            drawable = sAppIconCache.get(handle);
+            if (drawable == null) {
+                // if the drawable cached for this app is null don't bother trying to find it again
+                if (!sAppIconCache.containsKey(handle)) {
+                    new AsyncAppIconLoad(context, handle, callback).execute();
+                    return;
+                }
+            }
+        }
+        if (callback != null)
```

Comment:
You should move this to an else block:
```java
            drawable = sAppIconCache.get(handle);
            if (drawable == null) {
                // if the drawable cached for this app is null don't bother trying to find it again
                if (!sAppIconCache.containsKey(handle)) {
                    new AsyncAppIconLoad(context, handle, callback).execute();
                    return;
                }
            } else {
               <here>
            }
```

Reply:
I did not want the callback to be inside the sync block. I don't understand why you want it on the else branch.